alertmanager/alertmanager.go

```diff
@@ -0,0 +1,203 @@
+package alertmanager
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/prometheus/alertmanager/api"
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/dispatch"
+	"github.com/prometheus/alertmanager/inhibit"
+	"github.com/prometheus/alertmanager/nflog"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/provider/mem"
+	"github.com/prometheus/alertmanager/silence"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/alertmanager/ui"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/route"
+	"github.com/weaveworks/mesh"
+)
+
+// Config configures an Alertmanager.
+type Config struct {
+	UserID      string
+	DataDir     string
+	Logger      log.Logger
+	MeshRouter  *mesh.Router
+	Retention   time.Duration
+	ExternalURL *url.URL
+}
+
+// An Alertmanager manages the alerts for one user.
+type Alertmanager struct {
+	cfg        *Config
+	api        *api.API
+	logger     log.Logger
+	nflog      nflog.Log
+	silences   *silence.Silences
+	marker     types.Marker
+	alerts     *mem.Alerts
+	dispatcher *dispatch.Dispatcher
+	inhibitor  *inhibit.Inhibitor
+	stop       chan struct{}
+	wg         sync.WaitGroup
+	router     *route.Router
+}
+
+// New creates a new Alertmanager.
+func New(cfg *Config) (*Alertmanager, error) {
+	am := &Alertmanager{
+		cfg:    cfg,
+		logger: cfg.Logger.With("user", cfg.UserID),
+		stop:   make(chan struct{}),
+	}
+
+	am.wg.Add(1)
+	nflogID := fmt.Sprintf("nflog:%s", cfg.UserID)
+	var err error
+	am.nflog, err = nflog.New(
+		nflog.WithMesh(func(g mesh.Gossiper) mesh.Gossip {
+			return cfg.MeshRouter.NewGossip(nflogID, g)
+		}),
+		nflog.WithRetention(cfg.Retention),
+		nflog.WithSnapshot(filepath.Join(cfg.DataDir, nflogID)),
+		nflog.WithMaintenance(15*time.Minute, am.stop, am.wg.Done),
+		// TODO: Build a registry that can merge metrics from multiple users.
```

Comment:
I'll tag them `TODO(cortex)`, I'm fearing someone will wonder if `(mt)` is someone's username :)